### ISSUE TITLE
Dataset mean and variance

### ISSUE THREAD

**Issue by alex96295** [QiuJueqin/SqueezeDet-PyTorch#1]
Hello,

This is more like a question: why do we need to calculate the rgb.mean and rgb.std_dev of the dataset? And in your code, where are those information used? I only find the definition in the KITTI dataset configuration and no more.
Last:

- would it matter if I scale the input image dimensions by a factor 1/128?

- can I do training on the kitti dataset with model size 224x224? I mean by properly resizing input images and rescaling associated annotations

Thank you!

**Comment by QiuJueqin** [comment]
@alex96295 RGB mean and std are used for whitening the input image, as can be found here: https://github.com/QiuJueqin/SqueezeDet-PyTorch/blob/b0d80daa0147c968d43f3c38bc416394419c4bcc/src/datasets/base.py#L51

What does it mean that scale the input image dimensions by a factor 1/128?

Yes, you can resize input image to any size you like, but decreasing the resolution will also degrade the detection performance. In my exp rescaling KITTI by a factor 0.5 will cause a large mAP drop.

**Comment by alex96295** [comment]
@QiuJueqin thank you for the reply. I think that the 'scale' I am referring to is the std_dev, I mean images are pre-processed with mean 0 and std_dev = 128 to have output_im = (input_im - mean) / std_dev, where std_dev = 128 in my case

I mean, those are the mean and variance used for deployment on hardware. I am a little experienced with it, but I would ask: if I know that on the hardware, at inference time, I pre-process input images with std_dev = 128 and mean = 0, _on which mean and std_dev I should train the network_? On mean=0 and std_dev = 128 or by using the mean and the std_dev referred to the dataset? Does it have huge impact?

Ok thank you, I got it for the sizing and the resolution

I would add another question:

- what is the parameter forbid_resize? I have images in input which should be resized to 224x224 and the annotations, as well, should be adapted to the new dimensions. Should I set forbid_resize = false in my case, or go for the cropping and padding? What are the differences?

**Comment by QiuJueqin** [comment]
@alex96295 Thanks for clarification.

In this repo, I use RGB mean and std to match the pre-trained model from official PyTorch modelzoo. As pointed out [here](https://github.com/BichenWuUCB/squeezeDet/issues/87), training SqueezeDet from scratch is very difficult, so loading a pre-trained model and ensuring the input image having the same intensity range as those for training the pre-trained model are necessary.

I added the forbid_resize parameter to preventing the image texture been altered by the resizing operation in my application. You can ignore that and feel free to use resize function, i.e., do not add --forbid_resize option.

**Comment by alex96295** [comment]
@QiuJueqin ok thanks I got it. So if I happen to use the pre-trained model, I should use the mean and variance accordingly to match it. Are they generated from the code you provide in the readme? And if for inference I want to put mean = 0 and variance = 128, would it be feasible as well?

About the second point, I meant if resizing is equivalent to cropping and padding

**Comment by QiuJueqin** [comment]
Resizing is equivalent to cropping and padding only when the model's input size is identical to the image's original size. If the model's input size is much smaller than the original size, do not use crop&padding, otherwise most of image content will be cropped.

**Comment by alex96295** [comment]
@QiuJueqin OK got it, I have just edited the previous comment, please read also the second part

I thank you a lot :)

**Comment by QiuJueqin** [comment]
> 
> 
> @QiuJueqin ok thanks I got it. So if I happen to use the pre-trained model, I should use the mean and variance accordingly to match it. Are they generated from the code you provide in the readme? And if for inference I want to put mean = 0 and variance = 128, would it be feasible as well?
> 
> About the second point, I meant if resizing is equivalent to cropping and padding

Changing mean/std in train and inference will cause huge degradation. If using mean=0 and std=128 is crucial, try to train the model from scratch. If you find it difficult to converge, consider to train the SqueezeNet backbone first and then the SqueezeDet. Wish you luck!

**Comment by alex96295** [comment]
@QiuJueqin Ok, thanks you a lot!

**Comment by HassanBinHaroonHazen** [comment]
@QiuJueqin I need to train the SqueezeDet from scratch. Can you guide the important points for successful training from scratch? 

@alex96295 Have you successfully trained the SqueezeDet from scratch? kindly guide if yes, Thanks!

**Comment by QiuJueqin** [comment]
> @QiuJueqin I need to train the SqueezeDet from scratch. Can you guide the important points for successful training from scratch?
> 
> @alex96295 Have you successfully trained the SqueezeDet from scratch? kindly guide if yes, Thanks!

Unfortunately I couldn't train it from scratch in my own dataset. You can try it in some larger datasets, good luck.
